content/guides/nextjs/adding-tina/project-setup.md

```diff
@@ -0,0 +1,190 @@
+    return <ErrorPage statusCode={404} />
+  }
+
+  return (
+    //...
+    <PostBody content={post.content} />
+    //...
+  )
+}
+```
+
+The component accesses the HTML content of the post in `post.content`. We just saw how `getStaticProps` provides this content to the `Post` component. So, what we need to do here is add some code to transform the Markdown in `post.rawMarkdownBody` and send this to the `PostBody` component instead.
+
+First, add `useState` and `useEffect` to the imports at the top of this file:
+
+```js
+import { useState, useEffect } from 'react'
+```
+
+**In the body of the Post component**, let's create a State variable to store the transformed HTML. By initializing it with the already-transformed HTML, we will ensure that (1) the blog post will render immediately without having to wait for the initial transformation, and (2) the HTML will still be rendered during Next.js' _Server-Side Rendering_ step.
+
+```js
+export default function Post({ post, morePosts, preview }) {
+  //...
+
+  const [htmlContent, setHtmlContent] = useState(post.content)
+
+  return (
+    //...
+    <PostBody content={htmlContent} />
+    //...
+  )
+}
+```
+
+At this point, things should work identical to how they did before. Now, though, we can add our side effect to transform the raw Markdown and store it in our State variable:
+
+```jsx
+export default function Post({ post, morePosts, preview }) {
+  //...
+
+  const [htmlContent, setHtmlContent] = useState(post.content)
+  useEffect(() => {
+    markdownToHtml(post.rawMarkdownBody).then(setHtmlContent)
+  }, [post.rawMarkdownBody])
+
+  return (
+    //...
+    <PostBody content={htmlContent} />
+    //...
+  )
+}
+```
+
+Once you're done, you should be able to open up the `Post` component in React DevTools and edit the `rawMarkdownBody` to see the transformed HTML automatically updated in your browser.
+
+### 3. Prevent Unnecessary Transformation
+
+One minor downside of this is that, even though the Markdown is also transformed at build time, the client-side `markdownToHtml` call is always executed at least once. This may not seem like _that_ big of a deal, but since we're doing this exclusively to facilitate content editing, we should do everything possible to minimize the impact of Tina on visitors to your site.
+
+One way to improve this is to skip the client-side transformation step when the Post component is first mounted with the HTML generated at build time. To do this, let's add `useMemo` to our react imports at the top of the file:
+
+```js
+import { useState, useEffect, useMemo } from 'react'
+```
+
+We'll use `useMemo` to memoize the initial raw Markdown we receive from `getStaticProps`, and bail out of our `useEffect` if the current value of `rawMarkdownBody` hasn't changed from this initial value.
+
+```diff
+export default function Post({ post, morePosts, preview }) {
+  //...
+
+  const [htmlContent, setHtmlContent] = useState(post.content)
```

Comment:
I don't think we use `htmlContent`. 

Are we supposed to 
```
 - <PostBody content={post.content} />
 + <PostBody content={htmlContent} />
```

or something similar?

Reply:
are you looking at the latest version? both code examples that include the `htmlContent` declaration use it in `PostBody`

Reply:
You're right!